corehq/ex-submodules/toggle/tests.py

```diff
@@ -120,3 +128,118 @@ def test_toggle_cache(self):
 
         self.assertTrue(toggle_enabled(self.slug, 'mojer'))
         self.assertTrue(toggle_enabled(self.slug, 'fizbod', namespace=ns))
+
+
+class PredictablyRandomToggleSimpleTests(SimpleTestCase):
+
+    def setUp(self):
+        # Lie to get past settings.UNIT_TESTING check in PredictablyRandomToggle.enabled()
+        self.unit_testing = settings.UNIT_TESTING
+        settings.UNIT_TESTING = False
+
+    def tearDown(self):
+        settings.UNIT_TESTING = self.unit_testing
+
+    def test_deterministic(self):
+        self.assertEqual(
+            deterministic_random("[None, 'domain']:test_toggle:diana"),
+            0.5358778226236243
+        )
+
+    def test_all_namespaces_none_given(self):
+        toggle = PredictablyRandomToggle(
+            'test_toggle',
+            'A toggle for testing',
+            TAG_EXPERIMENTAL,
+            [NAMESPACE_USER, NAMESPACE_DOMAIN],
+            randomness=0.99
+        )
+        self.assertTrue(toggle.enabled('diana'))
+
+    def test_user_namespace_invalid(self):
+        toggle = PredictablyRandomToggle(
+            'test_toggle',
+            'A toggle for testing',
+            TAG_EXPERIMENTAL,
+            [NAMESPACE_USER],
+            randomness=0.99
+        )
+        with self.assertRaises(ValueError):
+            toggle.enabled('jessica')
+
+    def test_domain_namespace_invalid(self):
+        toggle = PredictablyRandomToggle(
+            'test_toggle',
+            'A toggle for testing',
+            TAG_EXPERIMENTAL,
+            [NAMESPACE_DOMAIN],
+            randomness=0.99
+        )
+        with self.assertRaises(ValueError):
+            toggle.enabled('marvel')
+
+    def test_user_namespace_enabled(self):
+        toggle = PredictablyRandomToggle(
+            'test_toggle',
+            'A toggle for testing',
+            TAG_EXPERIMENTAL,
+            [NAMESPACE_USER],
+            randomness=0.99
+        )
+        self.assertTrue(toggle.enabled('arthur', namespace=NAMESPACE_USER))
+
+    def test_domain_namespace_enabled(self):
+        toggle = PredictablyRandomToggle(
+            'test_toggle',
+            'A toggle for testing',
+            TAG_EXPERIMENTAL,
+            [NAMESPACE_DOMAIN],
+            randomness=0.99
+        )
+        self.assertTrue(toggle.enabled('dc', namespace=NAMESPACE_DOMAIN))
+
+
+class PredictablyRandomToggleTests(TestCase):
+
+    @classmethod
+    def setUpClass(cls):
+        cls.user_toggle = Toggle(
+            slug='user_toggle',
+            enabled_users=['arthur', 'diane'])
+        cls.user_toggle.save()
+        cls.domain_toggle = Toggle(
+            slug='domain_toggle',
+            enabled_users=[namespaced_item('dc', NAMESPACE_DOMAIN)])
+        cls.domain_toggle.save()
```

Comment:
I don't see these toggles being referenced anywhere. Are they needed?

Reply:
Well spotted.

:roll_eyes: Urgh. The toggles are not enabled for "jessica" and "marvel", not because they aren't members of `user_toggle` and `domain_toggle` (as this is supposed to test), but because "test_toggle" is not actually the slug of a persisted `Toggle` instance. ... i.e. They *should* be needed and the tests are passing for the wrong reason.